applications/matter_bridge/src/bridged_device_types/ble_environmental_data_provider.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "ble_environmental_data_provider.h"
+
+#include <bluetooth/gatt_dm.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+using namespace ::chip;
+using namespace ::chip::app;
+
+static bt_uuid *sServiceUuid = BT_UUID_ESS;
+static bt_uuid *sUuidTemperature = BT_UUID_TEMPERATURE;
+static bt_uuid *sUuidHumidity = BT_UUID_HUMIDITY;
+static bt_uuid *sUuidCcc = BT_UUID_GATT_CCC;
+
+bt_gatt_read_params BleEnvironmentalDataProvider::sHumidityReadParams{};
+
+BleEnvironmentalDataProvider *GetProvider(bt_conn *conn)
+{
+	BLEBridgedDevice *device = BLEConnectivityManager::Instance().FindBLEBridgedDevice(conn);
+	VerifyOrReturnValue(device, nullptr);
+	return reinterpret_cast<BleEnvironmentalDataProvider *>(device->mProvider);
+}
+
+bt_uuid *BleEnvironmentalDataProvider::GetServiceUuid()
+{
+	return sServiceUuid;
+}
+
+uint8_t BleEnvironmentalDataProvider::GattTemperatureNotifyCallback(bt_conn *conn, bt_gatt_subscribe_params *params,
+								    const void *data, uint16_t length)
+{
+	BleEnvironmentalDataProvider *provider = GetProvider(conn);
+	VerifyOrExit(provider, );
+	VerifyOrExit(data, );
+	VerifyOrExit(length == sizeof(mTemperatureValue), );
+
+	/* Save data received in notification. */
+	memcpy(&provider->mTemperatureValue, data, length);
+	DeviceLayer::PlatformMgr().ScheduleWork(NotifyTemperatureAttributeChange, reinterpret_cast<intptr_t>(provider));
+
+exit:
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+uint8_t BleEnvironmentalDataProvider::GattHumidityNotifyCallback(bt_conn *conn, bt_gatt_subscribe_params *params,
+								 const void *data, uint16_t length)
+{
+	BleEnvironmentalDataProvider *provider = GetProvider(conn);
+	VerifyOrExit(provider, );
+	VerifyOrExit(data, );
+	VerifyOrExit(length == sizeof(mHumidityValue), );
+
+	/* Save data received in notification. */
+	memcpy(&provider->mHumidityValue, data, length);
+	DeviceLayer::PlatformMgr().ScheduleWork(NotifyHumidityAttributeChange, reinterpret_cast<intptr_t>(provider));
+
+exit:
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+void BleEnvironmentalDataProvider::Init()
+{
+	/* Configure subscription for the temperature characteristic */
+	mGattTemperatureSubscribeParams.ccc_handle = mCccTemperatureHandle;
+	mGattTemperatureSubscribeParams.value_handle = mTemperatureCharacteristicHandle;
+	mGattTemperatureSubscribeParams.value = BT_GATT_CCC_NOTIFY;
+	mGattTemperatureSubscribeParams.notify = BleEnvironmentalDataProvider::GattTemperatureNotifyCallback;
+
+	/* Configure subscription for the humidity characteristic */
+	mGattHumiditySubscribeParams.ccc_handle = mCccHumidityHandle;
```

Comment:
I wonder, why we want to keep `mGattHumiditySubscribeParams` in a class, when we know the subscription establishment will fail? Wouldn't be better to just configure read parameters and add a comment that subscriptions are not supported? If we would like to implement generic check if subscription is supported maybe we should add a helper method to the BLEConnectivityManager. In this class it doesn't make much sense to me, as no-one will be able to re-use it outside anyway, and here it looks like redundant check.

Reply:
discussed f2f